route-rs-runtime/src/element/utils/icmp.rs

```diff
@@ -0,0 +1,262 @@
+    pub fn new(src_ip: Ipv4Address, bad_source_addrs: Vec<Ipv4Address>) -> Icmpv4ErrorGenerator {
+        Icmpv4ErrorGenerator {
+            src_ip,
+            mtu: IPV4_MIN_MTU,
+            ttl: 64, // TODO: larger to ensure it reaches original host?
+            bad_source_addrs,
+        }
+    }
+
+    /// Create an ICMPv4 TTL exceeded error message. Return `None` if an ICMP error message should
+    /// not be generated, and the packet should be silently discarded instead
+    ///
+    /// # Arguments
+    ///
+    /// * `packet` - The offending packet, the contents of which will be copied into the returned
+    /// error packet
+    pub fn ttl_exceeded_error(
+        &self,
+        packet: &mut Ipv4Packet<Vec<u8>>,
+    ) -> Option<Ipv4Packet<Vec<u8>>> {
+        self.generic_error(
+            packet,
+            Icmpv4Message::TimeExceeded,
+            Icmpv4TimeExceeded::TtlExpired.into(),
+            &[0; 4],
+        )
+    }
+
+    // Common way of constructing and ICMPv4 error message, including the 8 byte ICMP header and
+    // as many bytes as possible of the offending IPv4 packet up to the configured MTU size
+    fn generic_error(
+        &self,
+        packet: &mut Ipv4Packet<Vec<u8>>,
+        msg_type: Icmpv4Message,
+        msg_code: u8,
+        _optional: &[u8; 4],
+    ) -> Option<Ipv4Packet<Vec<u8>>> {
+        if !self.should_generate_error(packet) {
+            return None;
+        }
+
+        assert!(
+            Icmpv4ErrorGenerator::is_error_type(msg_type),
+            "Provided ICMP message type must be an error type"
+        );
+
+        // TODO: this will need to vary if we add support for source routing
+        // The size of the IPv4 header plus the overhead for the ICMP header
+        let min_header_size = 20 + ICMP_HDR_LEN;
+
+        // How much of the offending packet will be copied into the ICMP error
+        let copied_payload_len = cmp::min(packet.total_len() as usize, self.mtu - min_header_size);
+
+        let error_pkt_repr = Ipv4Repr {
+            src_addr: self.src_ip,
+            dst_addr: packet.src_addr(),
+            protocol: IpProtocol::Icmp,
+            payload_len: copied_payload_len + ICMP_HDR_LEN,
+            hop_limit: self.ttl,
+        };
+
+        // Generate the layer 3 header
+        let buffer: Vec<u8> = vec![0x00; error_pkt_repr.buffer_len() + error_pkt_repr.payload_len];
+        let mut error_pkt = Ipv4Packet::new_unchecked(buffer);
+        error_pkt_repr.emit(&mut error_pkt, &ChecksumCapabilities::ignored());
+        error_pkt.set_dscp(packet.dscp());
+        error_pkt.fill_checksum();
+
+        // Generate the ICMP message
+        let mut icmp_error = Icmpv4Packet::new_unchecked(error_pkt.payload_mut());
+        icmp_error.set_msg_type(msg_type);
+        icmp_error.set_msg_code(msg_code);
+        // TODO: fill in bytes 5 - 8 with _optional data for ICMP errors that require it
```

Comment:
These bytes are needed for some ICMP error messages that aren't yet supported in smoltcp